### ISSUE TITLE
View layer overhaul

### ISSUE THREAD

**Issue by igorbernstein** [activeadmin/activeadmin#2689]
While working on #2679, I started thinking about the view system as whole & wanted to start a discussion about a possible overhaul.  This ticket describes a lot of ideas that can & should be split up into separate tasks, but I wanted to discuss these changes in a broader context of ActiveAdmin's view layer road map and whatever changes that get a :thumbsup: can be spun off into separate issues.

It seems to me that a lot of the complexity in FormBuilder stems from the fact that it mixes 2 unreleated concerns: adding dynamic fieldset functionality to formtastic and trying to support Arbre's dsl that magicly concatenates html behind the scenes. Which needlessly couples the ActiveAdmin FormBuilder to Arbre, complicates if not prevents the FormBuilder's use in plain erb templates and creates weird edge cases like https://github.com/gregbell/active_admin/blob/master/spec/unit/form_builder_spec.rb#L440.  It also forces context sensitive behavior into the FormBuilder's implementation...for example if `input` is called in an `inputs` block then we use the form buffer otherwise just return the html.  In my experience the delegator/decorator pattern would be a better fit to support the buffering behavior: expose the decorator in the dsl to buffer the returned html, but leave the FormBuilder's internal behavior unmodified.  Looking at arbre's implementation for standard rails forms, it seems like thats the intended usage: https://github.com/gregbell/arbre/blob/master/lib/arbre/rails/forms.rb#L33.  I think it would be cleaner to extend FormBuilderProxy for the Formtastic & ActiveAdmin's FormBuilder's apis. This would allow forms just to use Arbre's buffers and address #2542

Arbre is cool, but seems to accomplish way more than is needed in this project.  Arbre's goal seems to be construct DOM in memory and allow the developer to manipulate it in an object oriented manner. However, from what I've seen in the project the only part that's really needed of Arbre is the html building DSL that allows a developer to quickly define the html structure inline with the rest of admin definition.  The unused in-memory DOM representation more than doubles the memory usage required to service a request and complicates the implementation to having a seperate buffer per element.  I think a lot can be gained my dropping the in memory DOM representation: a single ActionView buffer can be passed from parent to child, lowering memory usage and allow us to use [streaming responses](http://api.rubyonrails.org/classes/ActionController/Streaming.html).

While thinking about the Arbre changes, I was trying to figure out a good upgrade path for people who might've depended on the old behavior.  And I thought that there is no technical reason why we can't support multiple template engines in the view blocks:

``` ruby
form(engine: :erector) do |f|
  h1.heading! "First Name"
  ....
end

form(engine: :erb) do |f|
<<-EOS
  <%= f.inputs :first_name, :last_name %>
EOS
end
```

I also ran across erector, which seems to very similar to Arbre but seems an older projector and supports streaming: http://erector.rubyforge.org/faq.html#whyuseerector.  Any idea why this project was ruled out 2 years ago?

So to summarize:
Do you:
a. agree with:
b. be interested accepting PRs for:

the following changes:
1. decoupling of form buffers from FormBuilder
2. using a single buffer streaming architecture for the view api
3. switching out Arbre to something seemingly more mature like erector
4. configureable view engine

**Comment by varyonic** [comment]
I think it is time to wrap up this great discussion.
1. For better or worse FormBuilder has been integrated with Arbre using proxies, and form_buffers has been removed.  There are still issues because of the way Formtastic buffers its own output, and has_many has never been proxied.  There have been discussions of simple_form and cocoon but no follow up.  I think this item has been addressed as far as can be expected.
2. I think streaming responses for Arbre is a great idea and I added an issue referring here. I started working on something a while ago but it has not been a priority.
3. Erector (now Fortitude?) is interesting, though a major change, and introducing new dependencies is always a concern.  I think a separate issue should be opened if anyone is serious about going further.
4. I agree with Sean on multiple engines being overkill beyond partials and not a priority.
